### ISSUE TITLE
LAMP-worker uses filesystem

### ISSUE THREAD

**Issue by avaidyam** [BIDMCDigitalPsychiatry/LAMP-platform#382]
LAMP-worker [currently uses the filesystem (code here)](https://github.com/BIDMCDigitalPsychiatry/LAMP-worker/blob/4cd0af4f73f195015000a867b85318fdd6356d62/src/app.ts#L159) but there should be zero usage of filesystem across any LAMP backend code as there is no guarantee that these files/folders will remain available across container restart or resets.

**Comment by Linoy339** [comment]
@avaidyam .Currently, we are writing to filesystem in LAMP-worker and taking the path to execute inside the container.

 Just to clarify: we must take zip file from tag via tag api(currently there) and run the base64 encoded file directly inside the docker container instance. 
Right?

**Comment by avaidyam** [comment]
Yes. You should not be using the filesystem at all. Stream the files to the container if needed instead.

**Comment by Linoy339** [comment]
Thanks @avaidyam . We will update that

**Comment by Linoy339** [comment]
As we said, the base64  encoded string (script) uploaded against a token contains a zip file. Again this zip file contain script.js(or py) and requirements.txt. With the present implemenation(writing to filesystem),we get each file and its content. 
Now, we are trying to do it without uploading to filesystem, which makes difficult to read the individual files inside.

Can we make some tweaks on script upload which we would like to implement in such as 
way that the base64 encoded string  itself contains the requirements and the script file.
for example the sample string uploaded would be :
type:js,requirements:node-fetch;jsencrypt,iVBORw0KGgoAAAANSUhEUgAAAAUAAAAAANS//8IBKE0DHxgljNBA000DSSSSSAO==

Here, type is the file type(js or py) and the semi-colon seperated strings inside requirements are the dependancy of the script file and the the one starting after requirements part corresponds to script file(base64_encoded string).
so whenever the LAMP-worker gets this string for a token, it can easily convert this whole string into corresponding requirements and the script file easily
We would like to go with this. Can you suggest?

**Comment by avaidyam** [comment]
The contents of the ZIP file should be provided directly to the Docker container - we do not want to "break out" the requirements.txt contents like this. Why does LAMP-worker need to know what the contents of the requirements.txt file are? It should not need to know this info.

**Comment by Linoy339** [comment]
Okay. But In that case, reading content inside encoded(base64) string becomes difficult . Do you have idea on how to read zip file data without uploading it? 
To clarify, let me give you an example: Lets assume script.zip is our file and inside there requirements.txt and scriptfile.js.
script.zip gets encoded(base64) and it becomes iDDgAAAAUAAAAAANS//8IBKE0DHxgljNBA000DSSSSSAO==.
From this, its difficult to read  individual files like requirement.txt and scriptfile.js and its content.
If its uploaded, we can easily do it using the library which we are using now(admZip). If not uploaded, it becomes difficulty to read each file and it content. Buffer method to read content will give the whole content and separating requirements.txt and scriptfile.js becomes complicated. 
If you have any thoughts, please let us know.

**Comment by avaidyam** [comment]
We do not want to, nor do we need to, read the contents of the ZIP file? Let me recap this format for posterity with further explanation. 

---

The `Tag`'s value should be in this format: `data:application/lamp-automation;language=python3;trigger=/participant/U1234/sensor_event/lamp.gps;base64,<some base64 text representing a zip file here>`

1. `data:`: This is the marker for a [Data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URIs), similar to `http:` or `https:`. In `http:`, for example, a URL looks like `http://www.example.com/index.html`; the URL scheme is `http:` and the URL location is `//www.example.com/index.html`. Here, the `//` indicates "this is an absolute path, not a relative path." 
1. `application/vnd-lamp-automation`: This is the [MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) representing the data in this URL. We've made up `lamp-automation` for ourselves, and `vnd-` is short for "third-party vendor provided," meaning, this is not an [official MIME type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types) as declared by the IANA. 
1. `language=`: Data URLs support arbitrary companion parameters, and we declare `language` and `trigger` as parameters we need to know for the automation code. 
1. `python3`: The format of the `language` parameter is just a set of possible strings; here we support `python3`, `javascript` (or `js`), and `rscript`. These are arbitrary strings that we can change/control in the `LAMP-worker` code to map to the proper Dockerfile/base container required to run the automation code. 
1. `/participant/U1234/sensor_event/lamp.gps`: The format of the `trigger` parameter is identical to the NATS topic name, except using `/` instead of `.` to match actual API endpoints (this is what it should also be in NATS, if possible). 
1. `base64`: This indicates the contents of this Data URL are base64-encoded. 
1. `<some data>`: This is the actual base64-encoded data; we require this to be a ZIP file containing. In the case of `python3` (the contents can differ depending on language), it contains:
    1. an `__init__.py` file with a main function. **[REQUIRED]**
    1. a `requirements.txt` file with `pip` statements for requirements to be installed. _[OPTIONAL]_
    1. any other supporting files/libraries as required by the code.

---

The `LAMP-worker` code should: 
1. Upon initialization, locate all Tags declared this way (specified above, value begins with `data:application/vnd-lamp-automation`).
1. Register the topic handlers with NATS that match the triggers specified by these automations.
1. Listen for NATS callbacks on these topics. If a callback is received:
    1. Create a new Docker container with the base image specified (i.e. `python3`). 
    1. Copy the base64 ZIP file into this container and unpack it in a standard location. 
    1. Invoke the script with the payload received from NATS.
    1. Wait for script completion OR automatic timeout (say, 5 minutes maximum).
        1. Once termination is required, tear down and **delete** the Docker container. 
        1. Record execution stats (trigger, running length, etc.) to a log. 

---

I realize that not all of these pieces are built yet - that is fine. Following this above specification **(and NO deviation from it)**, @Linoy339, please go ahead with any code changes required.

**Comment by Linoy339** [comment]
Thanks @avaidyam.  
We believe all pieces except the filesystem ignore are done in LAMP-worker as we discussed in https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150. 

> The `Tag`'s value should be in this format: `data:application/lamp-automation;language=python3;trigger=/participant/U1234/sensor_event/lamp.gps;base64,<some base64 text representing a zip file here>`
This can be done. 



The changes would be :
1) change tag value  as mentioned above and key in tag would be like `study.cadn2efwhrxkppq9tn9t.activity.pqhm5zd2rbgpt5bjx5by`(its already there)

2) Instead of filesystem, we should copy whole string file(tag value) into new docker container and from there, we should unpack it and process. 
**Here I have some doubts: 
We need to run the contanier based on the language(py or js)used in tag value. Right? 
If so, if python, it would be python3 image or else it will node:alpine. Is that correct ?**
Also, the image used for js should contain npm or yarn, because installing these while script execution becomes time consuming and might results in memory issue.Can you confirm on these things?
 Also a suggestion: Instead of pulling a base image,can we have our own images with all these (JS,python,npm,pip etc)installed?

3)Both ways of script completion as well as automatic timeout has been done in LAMP-worker(https://github.com/BIDMCDigitalPsychiatry/LAMP-worker/blob/master/src/helpers/ScriptRunner.ts#L252).
We can implement accordingly.

4) Also, In our impelemntation, we  find the tag for a token when it gets published to nats. For example if `study.cadn2efwhrxkppq9tn9t.activity.pqhm5zd2rbgpt5bjx5by`  is the token which we get in LAMP-worker, we fetch the tag api with the attachment key as `study.cadn2efwhrxkppq9tn9t.activity.pqhm5zd2rbgpt5bjx5by` , so that we get the script file(base 64 encoded) in tag value as mentioned above(data:application/lamp-automation;language=python3;trigger=/participant/U1234/sensor_event/lamp.gps;base64,<some base64 text representing a zip file here>).
Please confirm.

5) In our implementation, only packages (py or js ) will be listed and not  any statements in requirements.txt. If py, we use pipe and if js, we use npm.
That would be fine. Right?

**Comment by avaidyam** [comment]
> We need to run the contanier based on the language(py or js)used in tag value. Right?

Correct.

> If so, if python, it would be python3 image or else it will node:alpine. Is that correct ?

Correct.

> Also, the image used for js should contain npm or yarn, because installing these while script execution becomes time consuming and might results in memory issue.Can you confirm on these things? Also a suggestion: Instead of pulling a base image,can we have our own images with all these (JS,python,npm,pip etc)installed?

Correct, and this makes sense. 

> Also, In our impelemntation, we find the tag for a token when it gets published to nats.

This implementation is the _reverse_ of what we should do. We need to locate all the Automations first, and set up ONLY the listeners required to handle them. To minimize code change at this point, upon worker start, build a dictionary mapping of these listener tokens mapped to the automation itself. 

> In our implementation, only packages (py or js ) will be listed and not any statements in requirements.txt. If py, we use pipe and if js, we use npm. That would be fine. Right?

This is not correct. [`requirements.txt` is a Python (`pip`, specifically) format](https://pip.pypa.io/en/stable/user_guide/) to list what dependencies the code has, you should use `pip install -r requirements.txt`. JS code has `package.json` and should use `npm install` there. We should either have one Dockerfile containing support for all supported languages or an individual Dockerfile for each supported language.

If you need examples, there are existing system like [Fx](https://github.com/metrue/fx) or [OpenFaaS](https://www.openfaas.com) that do this already.

**Comment by Linoy339** [comment]
@avaidyam . Thanks for the explanation and example.
With the discussion which we made here [https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150#issuecomment-845844999](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150#issuecomment-845844999), Can you please
brief which you said here : `This implementation is the reverse of what we should do. We need to locate all the Automations first, and set up ONLY the listeners required to handle them. To minimize code change at this point, upon worker start, build a dictionary mapping of these listener tokens mapped to the automation itself.`

When LAMP-worker starts and if we need to locate all the Automations tags, which tag api can be used ?
Again, I need you look at this thread([https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150#issuecomment-845958052](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/150#issuecomment-845958052)) which we already discussed on an issue if we save key as 'lamp.automation'.

**Comment by avaidyam** [comment]
Right, I think a good compromise right now is: 
1. Get the list of all Researchers 
2. For each, check if Tag "lamp.automation" exists. 
3. If a new Tag is uploaded the Worker should be getting a NATS message about it - check if it is a "lamp.automation" key. 
4. If a "lamp.automation" tag is created, updated, or deleted, update our internal mapping. 

This is a simple solution that will work for now. In the future we will want a more complex system instead but that will be too much work to implement for this beta release of Automations.

**Comment by Linoy339** [comment]
Okay. So the internal mapping would be something like this:
{researcherid_1:['study.123.activity'.*', 'study.123.activity'.124'], researcherid_2:['study.345.participant'.*', 'study.123.activity'.567']........} 
When a researcher uploads a script for particular token, this will be know by LAMP-worker via nats and add to this list
Is this right?
Also, do we also need to store the whole script in this internal  mapping, which seems not a good idea



> 2. For each, check if Tag "lamp.automation" exists: Can you please tell if it exists

**Comment by avaidyam** [comment]
You want the opposite mapping: the NATS token should map to an array of Researcher IDs. This way you know where the automation code is located for when the NATS message is received. (If the "lamp.automation" tag exists, we add the researcher ID to the mapping we created.)

We don't need to store the base64 data in memory, that is not a good idea as you have pointed out correctly.

**Comment by Linoy339** [comment]
Hi @avaidyam 
Can you please answer this:
If a particular researcher 'R1' has  uploaded script for 2 tokens study.123.activity.* and  study.456.activity.2. Also the script uploaded for both tokens are script1<base64 encoded> and script2<base64 encoded>

How would it appear in tag database?

Sample answer expected: 
{
_id:60a765103a2802fcfcd1e1e7,
_parent:"R1",
key:"lamp.automation",
type:"me",
_deleted:false,
value:"[
data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;base64;script1,
data:application/lamp-automation;language=js;trigger=/study/456/activity/2;base64;script2]"
}
Thanks

**Comment by avaidyam** [comment]
> How would it appear in tag database? 

This is not actually allowed or possible. A researcher-level user should not be able to create more than one `Tag` with the same key (here, `lamp.automation`). The `lamp.automation` key specifically needs to be a **STRING** in `data-url` format (as you have here). **We do not need to implement support for multiple scripts with different triggers/languages at this time.**

---

**For future reference only**, if there is a need to subscribe to multiple topics (i.e. multiple triggers), we can support multiple `trigger` attributes (i.e. `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;trigger=/study/456/activity/2;base64,...` or some other syntax (i.e. `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*&/study/456/activity/2;base64,...`). These will run into the issue of not being able to mix and match different languages (python/js, in your example), which is fine as a limitation for now.  

---

Also, [please note that according to the data-url syntax](https://en.wikipedia.org/wiki/Data_URI_scheme#Syntax), `base64;` should instead be `base64,` (with a **COMMA**, not a semicolon). 

Further, please try to use correct formatting syntax on GitHub issues (like ` for inline code or ``` for code blocks). I've edited your comment for readability.

**Comment by Linoy339** [comment]
@avaidyam  . Sorry for my syntax.

So, you are saying that the Data-URL itself is the key here:
`{
_id:60a765103a2802fcfcd1e1e7,
_parent:"R1",
key:"data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;base64",
type:"me",
_deleted:false,
value:"[
script1]"
}`

`{
_id:80a7652103a2802fcfcd1e1e7,
_parent:"R1",
key:"data:application/lamp-automation;language=js;trigger=/study/456/activity/2;base64",
type:"me",
_deleted:false,
value:"[
script2]"
}`
We need to get all the tags of this researcher when LAMP-worker starts (api : https://api.lamp.digital/type/R1/attachment) and if matches `lamp-automation` in the keys retrieved, pick it and push to our map array. Is that correct?
Can you confirm ?

**Comment by avaidyam** [comment]
> So, you are saying that the Data-URL itself is the key here:

No, I apologize, the `key` is still `lamp.automation`, the value is the data-url string, in the format `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;base64,<some base64 zip file here containing code and such>`. (There should not be `[` or `]`around the base64 zip file string.)

> We need to get all the tags of this researcher when LAMP-worker starts and if matches `lamp-automation` in the keys retrieved, pick it and push to our map array. Is that correct?

Yes, this part is correct. This needs to be done for ALL non-deleted researchers.

**Comment by Linoy339** [comment]
> Yes, this part is correct. This needs to be done for ALL non-deleted researchers.

That is understood.

So the value itself  assumes to contain multiple scripts(i.e. `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;base64,script1|
language=js;trigger=/study/456/activity/2;base64, script2`- something like this). Right?

**Comment by avaidyam** [comment]
No. I have specifically stated in my prior comment: 

> **We do not need to implement support for multiple scripts with different triggers/languages at this time.**

**Comment by Linoy339** [comment]
Okay.
So, what if a particular researcher have 2 scripts for 2 different activities(or study or sensor...) respectively?

**Comment by avaidyam** [comment]
Let's go with this: 
>  [...] if there is a need to subscribe to multiple topics (i.e. multiple triggers), we can support multiple trigger attributes (i.e. `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;trigger=/study/456/activity/2;base64,...`)

It should also be possible to use `trigger=/study/*/activity/*` - is this implemented?

**Comment by Linoy339** [comment]
> It should also be possible to use `trigger=/study/*/activity/*` - is this implemented?
No. Right now, key is the token and value is the script based on that particular key, So, even if the researcher has multiple scripts, its easy for them to upload with new token. The structure based on the present  implementation  would be:

`{
_id:60a765103a2802fcfcd1e1e7,
_parent:"R1",
key:"study.123.activity.*",
type:"me",
_deleted:false,
value:""script1..."//base 64 script
}`,
`{
_id:80a765103a2802fcfcd1e1e7,
_parent:"R1",
key:"study.456.activity.2",
type:"me",
_deleted:false,
value:""script2..." //base 64 script
}`


> [...] if there is a need to subscribe to multiple topics (i.e. multiple triggers), we can support multiple trigger attributes (i.e. `data:application/lamp-automation;language=python3;trigger=/study/123/activity/*;trigger=/study/456/activity/2;base64,...`)
My question is:
Each trigger would have different scripts and this should be included in the value itself. (i.e. `trigger=/study/123/activity/*base64, script1;trigger=/study/456/activity/2;base64, script2`. Right? If yes, the value would be higher in size for a researcher if lot of scripts are present for the same.

**Comment by avaidyam** [comment]
> No.

Let's implement proper wildcarding (i.e. `trigger=/study/*/activity/*`) as a priority. 

> The structure based on the present implementation would be: 

Right, so we want to move away from this to the implementation I shared in prior comments.  

> Right? If yes, the value would be higher in size for a researcher if lot of scripts are present for the same.

No - we do not want to support this. Please note in my example that it is one script/data-url that has multiple triggers. We should allow one script to have multiple triggers but not multiple scripts.

**Comment by Linoy339** [comment]
Okay.
So In the script/data-url, it should be a zip file even  for multiple triggers. Right? 

If yes, there would be multiple scripts inside the script file?(ie...if we unzip it, we get multiple folders, say study_456_activity_2  and  study_123_activity_*)

If yes, how would be invoke only one particular script (eg:study/456/activity/2 )

**Comment by avaidyam** [comment]
> So In the script/data-url, it should be a zip file even for multiple triggers. Right?

Yes. 

> If yes, there would be multiple scripts inside the script file?

No. 

> If yes, how would be invoke only one particular script

It is not possible. There is only one script that needs to be run when **any** of its triggers are activated. The script needs to be informed with what trigger activated the script.

**Comment by Linoy339** [comment]
Okay. Now, Understood.
So, we can assume there would be something like a switch case inside the script to run particular trigger(eg:study/*/activity/*,study/123/activity/456...) which we don't need to take care of .

We just need to give the token to the script and the script will do the rest. Is that Right?

**Comment by avaidyam** [comment]
Yes. You can pass the trigger that invoked the script to the script as an environment variable (i.e. `LAMP_TRIGGER`).

**Comment by avaidyam** [comment]
@Linoy339 It needs to be a ZIP file as it could be multiple script files or libraries. We cannot add the requirements.txt file to the data-url.

**Comment by Linoy339** [comment]
Okay. Understood
I have tweaked data url in a small way inorder to include the main script file :
`data:application/vnd-lamp-automation;driverscript=SampleScript.js;language=js;trigger=activity.ey0y0hdbrvwcxzvhqwfr.participant.*;base64,UEsDBBQAAAAIAJWWTFN1XajdDAAAAAoAAAAQAA`

Here, I have added a PARAM called `driverscript ` which will act as main script file to be executed. All others remains unchanged.

**Comment by Linoy339** [comment]
@avaidyam. Please find [https://github.com/BIDMCDigitalPsychiatry/LAMP-worker/pull/26/files](https://github.com/BIDMCDigitalPsychiatry/LAMP-worker/pull/26/files). We have tested it in our local environment and found to be working
We have merged it and need to test in staging.

**Comment by Linoy339** [comment]
@avaidyam.Also, we need to mount the volume(/var/run/docker.sock) in the staging  stack to make it work ?
Can we do it ?

**Comment by avaidyam** [comment]
@Linoy339 Thanks. Minor fix: [we already confirmed the format of the trigger string should use `/` and **NOT** `.`](https://github.com/BIDMCDigitalPsychiatry/LAMP-platform/issues/382#issuecomment-933513982) and that the way to add multiple triggers is to use `trigger=/activity/123;trigger=/activity/456` (please take note of the `/` in the beginning of the string and the repetition of `trigger=` with the `;` separator) instead of `&`. These details are minor but important - could you please fix those?

**Comment by Linoy339** [comment]
Okay. The issue is that we are unable to connect to the docker daemon (/var/run/docker.sock:ro) 
I did not understand why this is happening. Please check the permission/privileges .

**Comment by Linoy339** [comment]
@lukeoftheshire . Please find this link :
[https://phoenixnap.com/kb/cannot-connect-to-the-docker-daemon-error](https://phoenixnap.com/kb/cannot-connect-to-the-docker-daemon-error) which might help

**Comment by lukeoftheshire** [comment]
Hi @Linoy339 glad to hear it's working! To be honest I didn't get the chance to change anything after our previous discussion so I'm not sure what resolved the issue...perhaps just time. But thank you for the article - it will be a good resource for the future!

One note - while containers are being spun up, they are quickly failing and being removed with error 137 in the same way that the lamp-worker was initially failing (a couple blank terminations and then an error 137)- it's tough to access the logs as the containers are removed immediately, or I would give any logging info.

Let me know what other info I might give that would be helpful

**Comment by Linoy339** [comment]
@lukeoftheshire. Can you please check portainer, as we are unable to access the stack or containers. It says 
`Unable to retrieve stacks`.
Also can you please give the whole container logs for last 2 days?

**Comment by Linoy339** [comment]
Thank you @lukeoftheshire

**Comment by Linoy339** [comment]
@SuruPa00 . We have tested it weeks before and found working. We were also  waiting for your team's feedback and move to production if everything goes well. Hope @avaidyam can give us a feedback and move to production this week itself

**Comment by ZCOEngineer** [comment]
@avaidyam do you have any feedback on this? from our side we believe this is complete.

**Comment by avaidyam** [comment]
@ZCOEngineer Our team (me + @lukeoftheshire) will take a closer look later this month and double check things. This issue has been moved to a lower priority for now.

**Comment by ertjlane** [comment]
@avaidyam Should this issue be placed on hold, or do we think we should try to wrap this up?
